### PR TITLE
pool: built-in pools must be replicated

### PR DESCRIPTION
**Pull request by BlaineEXE**
**Description of your changes:**

Built in Ceph pools (device_health_metrics, .mgr, .nfs) must be
replicated and not erasure coded.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/apis/ceph.rook.io/v1/pool.go [hunk 1]
```diff
@@ -53,34 +53,47 @@ func (p *ReplicatedSpec) IsTargetRatioEnabled() bool {
...
+func ValidateCephBlockPool(s *NamedBlockPoolSpec) error {
+	name := s.Name
+	if name == "device_health_metrics" || name == ".mgr" || name == ".nfs" {
+		if s.IsErasureCoded() {
+			return errors.Errorf("invalid CephBlockPool spec: ceph built-in pool %q cannot be erasure coded", name)
+		}
+	}
+
+	return validatePoolSpec(s.ToNamedPoolSpec())
+}
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/pool.go:73]: I figured that we would only want to check the built-in pool names for CephBlockPools and not for named pools that might appear on file/object CRs.

File: pkg/apis/ceph.rook.io/v1/pool.go [hunk 2]
```diff
@@ -53,34 +53,47 @@ func (p *ReplicatedSpec) IsTargetRatioEnabled() bool {
...
+
+// validate any NamedPoolSpec
 func validatePoolSpec(ps NamedPoolSpec) error {
 	// Checks if either ErasureCoded or Replicated fields are set
 	if ps.ErasureCoded.CodingChunks <= 0 && ps.ErasureCoded.DataChunks <= 0 && ps.Replicated.TargetSizeRatio <= 0 && ps.Replicated.Size <= 0 {
-		return errors.New("invalid create: either of erasurecoded or replicated fields should be set")
+		return errors.New("invalid pool spec: either of erasurecoded or replicated fields should be set")
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/pool.go:79]: Updated the message to not assume the operation is a 'create' operation.

File: pkg/operator/ceph/pool/validate.go [hunk 3]
```diff
@@ -33,6 +34,11 @@ func validatePool(context *clusterd.Context, clusterInfo *client.ClusterInfo, cl
 	if p.Namespace == "" {
 		return errors.New("missing namespace")
 	}
+
+	if err := v1.ValidateCephBlockPool(&p.Spec); err != nil {
+		return err
+	}
+
```
Comment (BlaineEXE) [inline pkg/operator/ceph/pool/validate.go:41]: Now the CephBlockPool reconciler will use the same code as the admission controller for some of its validation.

**Comment by BlaineEXE** [comment]
I'm not sure how important this is to backport to v1.8. I expect some merge conflicts due to the `.mgr` pool not being present in v1.8.

File: pkg/apis/ceph.rook.io/v1/pool.go [hunk 4]
```diff
@@ -53,34 +53,47 @@ func (p *ReplicatedSpec) IsTargetRatioEnabled() bool {
...
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
+// ValidateCephBlockPool validates specifically a CephBlockPool's spec (not just any NamedPoolSpec)
+func ValidateCephBlockPool(s *NamedBlockPoolSpec) error {
+	name := s.Name
+	if name == "device_health_metrics" || name == ".mgr" || name == ".nfs" {
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/pool.go:66]: nit: why use a local var?
```suggestion
	if s.Name == "device_health_metrics" || s.Name == ".mgr" || s.Name == ".nfs" {
```

**Review by travisn — APPROVED** [review]
I'd vote not to backport to 1.8, but not a big deal either way. Approved with a small suggestion...

**Pull request merged**